### ISSUE TITLE
Site 053 Late Charges added to accounts that were paid

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1300]
In GitLab by @kdjstudios on Dec 31, 2018, 08:40

**Submitted by:**  "melissa.miller@answernet.com" <melissa.miller@answernet.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2018-12-31-45935

**Helpdesk Follow Up:** http://www.servicedesk.answernet.com/profiles/ticket/2019-01-15-36378/conversation

**Server:** Internal (All)

**Client/Site:** 053 (All)

**Account:** All

**Issue:** 

I have received a few calls that are on auto pay and were charged a late fee.  The payment was processed 12/11/18 and all account were changed to net 30.  Here are 2 examples so far.  I have already placed a stage fee credit to fix the accounts. The following accounts are TAS7573 Advance Publishing & E433 Dr Corbin. Please look into this matter I’m sure there are more.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 31, 2018, 08:43

@smasih @amishra @pchaudhary 

I reviewed both accounts provided above and it would appear that both were charged late fees incorrectly. Please look into this and provide an reason why they were charged late fees along with an estimate for a resolution to insure this does not continue to occur. Also find any other accounts that were charged late fees incorrectly.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 2, 2019, 01:30

@kdjstudios 

We dug into the code and found that first invoice "053-53422" under 053-TAS7573 ADVANCED PUBLISHING TECHNOLOGY account and second invoice "053-53464" under the 053-E433 DR. RICHARD CORBIN both invoices paid date are nil. 

While fixing of this ticket #1160 issue maybe that accounts were missed.

We have updated the paid date and if you regenerate the invoice it's working correctly.

Below invoices was paid date missing in a screenshot.
 
![image](/uploads/1295f175fc8139760b7fd5193c5b9c59/image.png)
![image](/uploads/71d9c9c71459cc3ae2ae73083b769040/image.png)

Hope this help.

thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 2, 2019, 09:05

@amishra 

Thank you for the update. A few things:

- What do you mean when you state regenerate the invoice? We are not able to regenerate a single invoice on single accounts, this would have to be done by your team. However, as stated in the ticket operations has already added the staged fees to correct this issue, but I would believe that if the late fee is not paid then another late fee will be added correct? I believe we will want to inform operations to generate manual invoices for each account to send the updated and correct balance to the client.  
- Did we check all accounts as requested? http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1303#note_32155
 - I will be opening a new ticket to check all other sites too. #1306


I wrote:

> The support team was able to find on both accounts an invoice that was paid but was showing historically that it was not paid. They have manually fixed this discrepancy and are going to check all other accounts to insure there are no other issues. We will follow up accordingly. 
> 
> We would recommend that for each account to manually create an invoice to reflect the updated account balance with the credit for the miss-charged late fees. This will insure the clients are aware of the correct balance to pay. Please let us know if you have any questions or concerns.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 2, 2019, 10:50

Melisa wrote:

> Here are a few more examples:
> 053T7664
> T8300
> Will I have to go through each account with a late fee to credit or will you be able to handle that?

I wrote:

> Thank you for the update. Please allow the support team to review all of the accounts at your site to find all that were charged incorrectly. Please hold off on going through the accounts manually. We should be able to have the support team add credits to each account that was charged a late fee in error and manually create an invoice with an updated balance accordingly. We will follow up with the list of accounts for you to then follow up with the clients accordingly so they are made aware of the correct balance. 
> 
> May you please confirm for the credits the support team should be using the "7882 - Late Charge Credit" code?

@smasih @amishra @pchaudhary @tim.traylor 

Since this was due to a programming error not a user error we will need to fix this. The following needs to occur:

1. Check all accounts for this site.
2. Confirm which accounts have invalid late fees assessed on them
3. Post a credit using "7882 - Late Charge Credit" to equal the value of the late fee.
4. Generate manual invoices for each account.
5. Provide a list of all accounts and the corrections made to each, so we may provide this to operations to notify their clients accordingly.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 3, 2019, 06:43

@kdjstudios 

Here is the list of all invoices which paid date are nil. So we need to update the paid date of the invoices.

Please confirm with the Melisa.

[53_invalid_account_invoice_date.csv](/uploads/3a94f6335fbe24dce9a337d79c8b4ceb/53_invalid_account_invoice_date.csv)

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 3, 2019, 09:28

@amishra 

Thank you for the list, I will reach out to Melissa for her approval. 

I wrote:

> The support team was able to compile a list of all accounts and their invoices that have an invalid paid date on them. If you could please review the list and provide your approval we will have the support team add the staged fees and create manual invoices for each accordingly. Note you will need to check each account to insure that the late fee is invalid. Please feel free to update the list with which accounts are approved to have a credit added to them. Please let us know if you have any questions or concerns. 

@amishra @pchaudhary @smasih @tim.traylor @abakshi 

I really hope this is a one of a kind site and the others are not this bad. If I see this correctly there are 35 accounts on the list. There is about 175 accounts overall at Atlanta. so that is that 20% of the accounts are being charged invalid late fees? Please see ticket #1306, I will be updating its' priority to insure we address this before more accounts are charged with invalid late fees.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 3, 2019, 11:16

@amishra 

Please use the provided spreadsheet to add staged fee credits to each account accordingly. The credit will need to be equal to the amount of late fees that were incorrectly assessed on each account accordingly. Once that is done, please let us know. I will then update Melissa to confirm that the accounts will be in good standing and we can manually draft invoices accordingly to insure the account balances are updated prior to the sites next billing cycle to insure more invalid late fees are not assessed.   

Mellisa wrote:

> I have attached the new spreadsheet. All are approved for a stage fee to be added. I have added the couple accounts that I already added the stage fee added.

> [53_invalid_account_invoice_paid_date__1_.xlsx](/uploads/2e4603f87526e17d80af12112883ce1f/53_invalid_account_invoice_paid_date__1_.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 4, 2019, 08:21

@kdjstudios 

This issue is not specific to Atlanta 

**This was existing an bug and got surfaced after release of #999 (we moved credit calculation from draft stage to final stage) because existing code was causing further issues**

Intensity of this issue was low because **paid_date** on invoice was not getting updated only when it was paid off using another credit invoice.

Here is the list of invoices (valid: means valid late charge) [account_with_late_fees106__copy_.csv](/uploads/db597acbc3ef6464a20a52b9d3f54563/account_with_late_fees106__copy_.csv)

**We have updated the code and paid date to stop this issue from happening further in future**

Only thing we are confused with is how to calculate due date

for eg

take the account **053-TAS7573** in Atlanta Site  
Invoice date: "01/01/2019"   
Due date as per code: "01/01/2019" - 30(net terms) = **12/02/2018**   
Payment made against full balance: "12/11/2018"

As per system calc, the payment was made after due date

but how could we expect someone to pay just next day after invoice

**In existing late fee code there is a bug which never considers payment date. Meaning invoice is paid off (before invoice generation) system will not charge late fee irrespective of the payment date.**

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 4, 2019, 09:58

@pchaudhary @amishra @smasih @pchaudhary @abakshi @rchauhan @tim.traylor 

Please do not use the current late fee calculation. We all already know it is incorrect. #1191 

So to my understanding we need to do the following:

1. Finish #1191 and release it.
2. Finish #1306 and release it.
3. Once the two above are done we can then revisit the underlining issue here about payments not setting paid dates?

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 5, 2019, 04:22

@kdjstudios 

Here is the list of those accounts of Atlanta which do have the late fees charged after the fix of the previous late fees issue.  
[Atlanta_late_accounts.xlsx](/uploads/11c0c4138ec96c242850c45bb294db73/Atlanta_late_accounts.xlsx)

The above list contains all of the late fees charged accounts & there is a column named *late_fee_is_valid?* which determines the charged late fees is valid or not? 

The above list also contains the accounts on which you mentioned that the staged fee was already placed. (under the invalid charged category)  

* We need the permission to fix the accounts by adding the credit staged fees of the same amount.  
*  Although, we guess that it should be *Late Charge Credit* **7882** but still we need the confirmation on the code of the staged fee which should be added while adding the staged fees.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 5, 2019, 04:26

@kdjstudios 

I understand that you might have the question about the previous list but we need to clarify that the previous list does contain accounts pulled by the basis of paid date only however we need the accounts which incorrectly included the late charges.  

We also would like to confirm that this new list contains the accounts of the previous list also.

Hope this helps
Thanks

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jan 7, 2019, 10:59

@smasih @rchauhan I am totally beside myself on this issue.  We spent tens of hours (probably over a hundred counting everybody's hours) identifying the underlying issue here, running fixes, and defining what the late charge processing is actually supposed to be doing. If I'm understanding correctly, the root problem (that had been hot fixed before we ran all of the programmatic adjustments) was somehow 'lost' or overwritten and now we're back to months and months of incorrect late charges that will now have to be identified and corrected?

If this is what happened then it is a huge failure that we need to discuss tomorrow.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 9, 2019, 09:00

@kdjstudios 

We have again researched about the root cause for the issue as suggested by Tim on yesterday's call and we found exactly what we were saying at early.

* We dumped the previous DB on merge & checked for the account *DR. RICHARD CORBIN*.
* There is an invoice *053-53464* having paid date as nil which was causing the issue.
* The reason of **Nil** paid date is the previous bug on the SAB.
  * If there is an invoice having the credit amount (other than payment) which is canceling the invoice balance of any previous invoice then the balance of the previous invoice will calculate correct as zero but the paid date will not be updated on the previous invoice.

* The issue surfaced up when we added the credit late fees for the accounts which were canceling the previous invoice balance but did not update the paid date on the invoice.

* We have already updated the paid date for all of the invoices lying on such case. 

* We have fixed the underline issues on both servers.  

* Answernet - [answernet_account_with_late_fees.csv](/uploads/130b1dbc3380c8972947d962fc62e22f/answernet_account_with_late_fees.csv)

* Hosted - [hosted_account_with_late_fees.csv](/uploads/81f65719eda637a0005ecf2d1a046797/hosted_account_with_late_fees.csv)

* Hosted has 131 accounts & out of which 14 are with the net terms greater than 30, so we have to check those accounts manually.

* Answernet has 247 accounts & out of which 27 are with the net terms greater than 30, so we have to check those accounts manually.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 9, 2019, 12:43

@pchaudhary @amishra @smasih 

Please review this with @tim.traylor in your morning meeting to find the appropriate plan of action.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 11, 2019, 06:11

@kdjstudios 

We didn't have the discussion with Tim around the plan of action. 
Is there something you can suggest?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 11, 2019, 08:55

@pchaudhary

Please review this with @tim.traylor in your morning meeting to find the appropriate plan of action.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jan 14, 2019, 06:11

As per our discussion, I now understand that it was the fixing of the previous late charge overcharges that created a scenario where we had so many credit invoices, and that in turn cause the missed invoice date update to impact so many accounts.

The Aavaz team will move forward on reversing the bad late charges the same way we did previously.  A new list will be provided and Ops needs to approve that we'll credit the accounts.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 14, 2019, 06:55

@kdjstudios  @tim.traylor 

Please find the New list of the accounts-

 [answernet_account_with_late_fees_latest.csv](/uploads/af6f5398dae3c1c8ecf34d56b51776e0/answernet_account_with_late_fees_latest.csv)

[keener_account_with_late_fees_latest.csv](/uploads/618fc190f892bb53be7b5e87855dd724/keener_account_with_late_fees_latest.csv)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 15, 2019, 08:50

Mellisa wrote:

> Is there an update on the stage fees being added?  Will I have to process the stage fees?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 15, 2019, 09:18

@tim.traylor wrote an email to Gary to update him. I am writing an email for approval by Terri before sending to ops and clients. Once we have approval then we will follow up accordingly.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 15, 2019, 10:59

I wrote to Melissa:

> The support team is actually reassessing the list of valid vs invalid late fees. This bug actually effected multiple sites and we anticipate to have an email with updated list of accounts to all sites that were effected within a day requesting approval to add credits to each sites accounts effected accordingly.
> 
> The support team will be able to add the credits to the accounts.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 15, 2019, 15:23

After informing Gary and receiving approval on the email wording. I was able to send out the emails to all sites and clients whom had invalid late fees on their list. I will be posting a new gitlab ticket for each site client as we receive feedback from them accordingly.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 17, 2019, 05:25

@kdjstudios 

We have successfully added the staged fee on **Allentown, PA** site.   
Please find the below attachment for the details.

[allentown_added.csv](/uploads/1c9257432fe673604d22384118ae6970/allentown_added.csv)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 17, 2019, 16:21

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 18, 2019, 03:56

@kdjstudios @tim.traylor 

There is one another thing which we are going to include after your approval.

The late charges will be added to the next invoice in the case User has only submitted the valid invoice payment from the last one. 

* In this case, the remaining late charges balance is going to be forward & making an impact on the next invoice.
* While the creation of next invoice, the late charges are being calculated by looking into the previous balance which is still remaining & also equals to the previous invalid late charges.

Eg: 
  * Let's say there is an invoice I1 which is having the total balance of $125 from which the $25 is the invalid late charges.
  * User paid $100 only (the valid balance of the invoice)
  * Now we add $25 staged fees into the account.
  * Then while creating a new invoice, the existing code takes the previous remaining $25 as the remaining account balance & provoke the late fee based on that $25 which is logically wrong.
  * We need to stop the add the late charge in this scenario.
  

* We hope that Tim remember this scenario from last time.


Thanks

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jan 18, 2019, 06:37

@pchaudhary Hi Piyush, I remember us doing the additional collection that had the accounts that had been incorrectly late charged. We used that to make sure we did not late charge those same accounts if they had not paid the (invalid) late charge amount from the previous invoice.  We definitely want to do exactly the same thing this time.  Thx, Tim

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 18, 2019, 12:42

Melisa wrote:

> Here are the updated accounts with invalid late fees.  Please use code 7882 Late Charge Credit. 

[Atlanta+1-15-19.xlsx](/uploads/a3bc7e16820ffc3b624801a7daded384/Atlanta+1-15-19.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jan 19, 2019, 05:01

@kdjstudios 

We have successfully added the staged fee on **Atlanta** site.  
Please find the below attachment for the details.  
[atlanta_Added.csv](/uploads/4586a848e80a485c3b6eb0d8d5735b3e/atlanta_Added.csv)


thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 21, 2019, 09:00

I wrote:

> The support team was able to add credits to the following accounts on the attached spreadsheet. Please let us know if you have any questions or concerns.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 21, 2019, 09:02

@amishra @smasih @pchaudhary 

I just want to confirm what is the status of the bug in this ticket? Has it been resolved and released to production? Or have we only done the manual fixes on the accounts so far?

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 22, 2019, 01:07

@tim.traylor 

> Hi Piyush, I remember us doing the additional collection that had the accounts that had been incorrectly late charged. We used that to make sure we did not late charge those same accounts if they had not paid the (invalid) late charge amount from the previous invoice.  We definitely want to do exactly the same thing this time.  Thx, Tim

* Thanks for the update. We understand your concern & will want to ask some points:
  * The BadLateCharge code was written in such manner to support only one entry but recently we have found that there may be more than one entry of an account.  
  * It will not work now as there are many accounts which were already added on the last time while adding BADLateCharge entry.  
  * So, either, we have to delete the old entries or have to change the code. 

* We request you to mention your opinion on the same.

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 22, 2019, 02:37

@kdjstudios 

> I just want to confirm what is the status of the bug in this ticket? Has it been resolved and released to production? Or have we only done the manual fixes on the accounts so far?

* I understand your concern & want to state here that as already mentioned in [comment] (http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1303#note_32555) we have already fixed the bug spotted on this issue.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 22, 2019, 11:36

@pchaudhary 

Thank you for the update confirming that the underling bug has been resolved. Were you able to review the information in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1303#note_33260 with @tim.traylor in the meeting this morning? If not please review it with him during tomorrows morning meeting.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 23, 2019, 24:50

@kdjstudios @tim.traylor 

We would like to mention here that we need a clarification on a point where the late charges were charged more than once on an account but we have credited it only once. (as mentioned in #1341)

- **What should we do with such type of cases?**

- **Should we credit the late fee amount by the no of times it was invalid charged previously?**

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 23, 2019, 05:05

@kdjstudios @tim.traylor 

We also would like to mention a thing that as we previously stated on the [comment] (http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1303#note_32555), the accounts we noticed are having the net terms greater than 30.  

* So we have to manually look into those accounts.

Hope this helps.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 23, 2019, 06:32

@kdjstudios @smasih 

We have created the [Wiki] (http://gitlab.aavaz.biz/AnswerNet/SABilling/wikis/Late-Fee-Credit) for consolidating this ticket.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 23, 2019, 06:37

@kdjstudios 

Please find the list of the accounts having  [Ans_accounts_having_invalid_entries_more_than_once.csv](/uploads/bd16b2bff7789ee34fa507ba5375a89a/Ans_accounts_having_invalid_entries_more_than_once.csv) 
* As we have previously mentioned that for the accounts having net terms greater than 30, we have to manually check the validation of invalid late fees charged.  
* We looked into some accounts & tried to check the validation on those late charges but unable to find the correct way to check that.
* We need help to check.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 23, 2019, 09:09

@pchaudhary the link in comment http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1303#note_33410 to the accounts displays 404 error. please upload again.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 23, 2019, 09:10

We will need to review this with @tim.traylor and possibly Gary.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 23, 2019, 23:34

@kdjstudios 


> The link in comment http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1303#note_33410 to the accounts displays 404 error. please upload again.

* Uploaded. Please check

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jan 25, 2019, 09:38

Need to review with @tim.traylor to see confirm we just need to reach out to each of those site and ask for approval to add credits for previous invoices also.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jan 28, 2019, 05:20

@kdjstudios @tim.traylor 

Just for information & reminder that there are still 2 points which need to be taken care of:

* Where the late charges were charged more than once on an account but we have credited it only once.  
* For the accounts having net terms greater than 30, we have to manually check the validation of invalid late fees charged. http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1303#note_33410

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 7, 2019, 09:28

Reviewed with @tim.traylor and we believe this was already handled.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 14, 2019, 11:15

I wrote:

> We are just following up on this ticket. Over the last 6 months there has been multiple improvements and adjustments to how late fees are assessed and calculated. This updates have resolved this ticket and we will be closing the ticket now. If you have any questions or concerns please let us know.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 14, 2019, 11:19

closed
